ocaml/xapi/xapi.ml

```diff
@@ -783,6 +783,13 @@ let set_stunnel_timeout () =
     Stunnel.timeoutidle := Some timeout
   with _ -> debug "Using default stunnel timeout (usually 43200)"
 
+let init_tls_verification ~__context () =
+  (* set global flag based on database entry *)
+  let pool = Helpers.get_pool ~__context in
+  let default = Db.Pool.get_tls_verification_enabled ~__context ~self:pool in
```

Comment:
This is not right in the case where a user emergency disabled verification on a host with `Host.emergency_disable_tls_verification`

I think `Stunnel_client.set_verify_by_default` should persist the setting as a file flag, and then we don't even this function

Reply:
We need to be aware that the library is used by different binaries: at least xapi and xe, So who needs to maintain the state?

Reply:
Alternatively, is there a way to check whether we are in emergency mode and initialize the default correctly here?

Reply:
> Alternatively, is there a way to check whether we are in emergency mode and initialize the default correctly here?

on master right now, the check is just whether /var/xapi/verify-certificates exists or not (but with your changes that won't work anymore)

> We need to be aware that the library is used by different binaries: at least xapi and xe, So who needs to maintain the state?

It's clear that xapi should maintain at least its own state. Xapi could also maintain the state of xe clients which are running on hosts in a pool, but this might not be a good idea - e.g. what if you want to make an xe call to another pool? So you make a good point :+1: 

If we say that each binary should decide for itself how it wants to maintain its verification state, then having this `init_tls_verification` function does make sense (xe presumably needs one too?). Ultimately, for the xapi binary, the database can't be used for maintaining this state, so I would propose a file flag

Reply:
How about: pool-enable-tls-verifcation removes a file flag and host-emergency-disable-tls-verification creates one? On startup we check for the file.

Reply:
> How about: pool-enable-tls-verifcation removes a file flag and host-emergency-disable-tls-verification creates one? On startup we check for the file.

This would mean 'verification enabled if and only if Db.Pool.get_tls_verification_enabled && file doesn't exist'. I think it will have to be: 'file exists if and only if verification enabled' (because of the chicken and the egg problem where we would need to contact the database to decide whether to enable verification, but in order to read the database we have to open a connection, but we don't know whether it should be verified or not ...)

Reply:
For now I would suggest we merge this and know that emergency mode for TLS does not survive a toolstack restart. But we have to think about of the semantics of these emergency modes. I feel there is an asymmetry how it is enabled and disabled